File: library/src/main/java/com/yy/codex/uikit/UISwitch.java
package com.yy.codex.uikit;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Outline;
import android.os.Build;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.RequiresApi;
import android.util.AttributeSet;
import android.view.View;
import android.view.ViewOutlineProvider;

/**
 * Created by cuiminghui on 2017/1/17.
 */

public class UISwitch extends UIControl {

    public UISwitch(@NonNull Context context, @NonNull View view) {
        super(context, view);
    }

    public UISwitch(@NonNull Context context) {
        super(context);
    }

    public UISwitch(@NonNull Context context, @NonNull AttributeSet attrs) {
        super(context, attrs);
    }

    public UISwitch(@NonNull Context context, @NonNull AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
    public UISwitch(@NonNull Context context, @NonNull AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
    }

    private UIView mOffBackgroundView;
    private UIView mOnBackgroundView;
    private UIView mHandleView;
    private float mHandleRadius = 30; // realRdius_28 + shadowXY_2
    private boolean mOn;
    private boolean mActive;
    private UIViewAnimation mCurrentAnimation = null;

    // configuable
    private UIColor mOnThumbColor = UIColor.Companion.getWhiteColor();
    private UIColor mOnTrackColor = new UIColor(0x00/255.0, 0xe3/255.0, 0x64/255.0, 1);
    private @Nullable Bitmap mOnBitmap = null;
    private UIColor mOffThumbColor = UIColor.Companion.getWhiteColor();
    private UIColor mOffTrackColor = UIColor.Companion.getWhiteColor();
    private @Nullable Bitmap mOffBitmap = null;

    public void setOffTraceColor(UIColor mOffTraceColor) {
        this.mOffTrackColor = mOffTraceColor;
    }

    public void setOffThumbColor(UIColor mOffThumbColor) {
        this.mOffThumbColor = mOffThumbColor;
    }

    public void setOnThumbColor(UIColor mOnThumbColor) {
        this.mOnThumbColor = mOnThumbColor;
    }

    public void setOnTraceColor(UIColor mOnTraceColor) {
        this.mOnTrackColor = mOnTraceColor;
    }

    private void defaultValue(){
        mHandleRadius = 30;
        mOnThumbColor = UIColor.Companion.getWhiteColor();
        mOnTrackColor = new UIColor(0x00/255.0, 0xe3/255.0, 0x64/255.0, 1);
        mOnBitmap = null;
        mOffThumbColor = UIColor.Companion.getWhiteColor();
        mOffTrackColor = UIColor.Companion.getWhiteColor();
        mOffBitmap = null;
    }

    @Override
    protected void init() {
        super.init();
        defaultValue();
        mOffBackgroundView = new UIView(getContext());
        mOffBackgroundView.setFrame(new CGRect(0, 0, 51, 32));
        mOffBackgroundView.setAlpha(1.0f);
        mOffBackgroundView.setWantsLayer(true);
        mOffBackgroundView.getLayer().setCornerRadius(16);
        mOffBackgroundView.getLayer().setBorderWidth(2);
        mOffBackgroundView.getLayer().setBorderColor(new UIColor(0x00/255.0, 0x00/255.0, 0x00/255.0, 0.1));
        mOffBackgroundView.getLayer().setBackgroundColor(mOffTrackColor);

        mOnBackgroundView = new UIView(getContext());
        mOnBackgroundView.setFrame(new CGRect(0, 0, 51, 32));
        mOnBackgroundView.setAlpha(0.0f);
        mOnBackgroundView.setWantsLayer(true);
        mOnBackgroundView.getLayer().setCornerRadius(16);
        mOnBackgroundView.getLayer().setBackgroundColor(mOnTrackColor);

        mHandleView = new UIView(getContext());
        mHandleView.setFrame(new CGRect(2, 2, mHandleRadius, mHandleRadius));
        mHandleView.setAlpha(1.0f);
        mHandleView.setWantsLayer(true);
        mHandleView.getLayer().setShadowX(2);
        mHandleView.getLayer().setShadowY(2);
        mHandleView.getLayer().setShadowRadius(0.5);
        mHandleView.getLayer().setShadowColor(new UIColor(.3, .3, .3, .2));
        mHandleView.getLayer().setCornerRadius(14);
        mHandleView.getLayer().setBorderWidth(0.5);
        mHandleView.getLayer().setBorderColor(new UIColor(0x00/255.0, 0x00/255.0, 0x00/255.0, 0.15));
        mHandleView.getLayer().setBackgroundColor(mOffThumbColor);

        addSubview(mOffBackgroundView);
        addSubview(mOnBackgroundView);
        addSubview(mHandleView);
    }

    @Override
    public void layoutSubviews() {
        super.layoutSubviews();
    }


    @Override
    protected void onEvent(Event event) {
        super.onEvent(event);
        switch (event){
            case TouchDown:
//                NSLog.warn("---");
//                NSLog.warn("touch down");
                if (!mActive){
                    setActiveAnimated(true);
                }
                break;
            case TouchDragExit:
//                NSLog.warn("drag exit");
                setOnAnimated(this.mOn);
                break;
            case TouchUpInside:
//                NSLog.warn("touch upInside");
                setOnAnimated(!this.mOn);
                break;
            case TouchUpOutside:
//                NSLog.warn("touch upOutside");
                setOnAnimated(this.mOn);
                break;
        }
    }

    private void setActiveAnimated(final boolean isActive){
        this.mActive = isActive;
        cancelAnimation();
        mCurrentAnimation = UIViewAnimator.INSTANCE.spring(new Runnable() {
            @Override
            public void run() {
                float widthExpanded = 8;
                if (isActive) {
                    if (mOn) {
                        mHandleView.setFrame(new CGRect(21 - widthExpanded, 2, mHandleRadius + widthExpanded, mHandleRadius));
                    } else {
                        mHandleView.setFrame(new CGRect(2, 2, mHandleRadius + widthExpanded, mHandleRadius));
                        mOffBackgroundView.getLayer().setBackgroundColor(new UIColor(0x00 / 255.0, 0x00 / 255.0, 0x00 / 255.0, 0.1)); // darken +30%
                    }
                } else {
                    if (mOn) {
                        mHandleView.setFrame(new CGRect(21, 2, mHandleRadius, mHandleRadius));
                    } else {
                        mHandleView.setFrame(new CGRect(2, 2, mHandleRadius, mHandleRadius));
                        mOffBackgroundView.getLayer().setBackgroundColor(mOffTrackColor);
                    }
                }
            }
        }, new Runnable() {
            @Override
            public void run() {
                mActive = false;
            }
        });
    }

    public void setOnAnimated(boolean isOn){
        this.mOn = isOn;
        cancelAnimation();
        mCurrentAnimation = UIViewAnimator.INSTANCE.spring(new Runnable() {
            @Override
            public void run() {
                if (mOn) {
                    mOnBackgroundView.setAlpha(1.0f);
                    mOnBackgroundView.getLayer().setBackgroundColor(mOnTrackColor);
                    mOffBackgroundView.setAlpha(0.0f);
                    mHandleView.setFrame(new CGRect(21, 2, mHandleRadius, mHandleRadius));
                    mHandleView.getLayer().setBackgroundColor(mOnThumbColor);
                } else {
                    mOffBackgroundView.setAlpha(1.0f);
                    mOffBackgroundView.getLayer().setBackgroundColor(mOffTrackColor);
                    mOnBackgroundView.setAlpha(0.0f);
                    mHandleView.setFrame(new CGRect(2, 2, mHandleRadius, mHandleRadius));
                    mHandleView.getLayer().setBackgroundColor(mOffThumbColor);
                }
            }
        }, new Runnable() {
            @Override
            public void run() {
                mActive = false;
            }
        });
    }

    private void cancelAnimation(){
        if (mCurrentAnimation != null){
            mCurrentAnimation.cancel();
        }
    }

    public void setOn(boolean isOn){
        if (this.mOn == isOn) {
            return;
        }
        this.mOn = isOn;
        if (isOn){
            mOnBackgroundView.setAlpha(1.0f);
            mOnBackgroundView.getLayer().setBackgroundColor(mOnTrackColor);
            mOffBackgroundView.setAlpha(0.0f);
            mHandleView.setFrame(new CGRect(21, 2, mHandleRadius, mHandleRadius));
            mHandleView.getLayer().setBackgroundColor(mOnThumbColor);
        }
        else {
            mOffBackgroundView.setAlpha(1.0f);
            mOffBackgroundView.getLayer().setBackgroundColor(mOffTrackColor);
            mOnBackgroundView.setAlpha(0.0f);
            mHandleView.setFrame(new CGRect(2, 2, mHandleRadius, mHandleRadius));
            mHandleView.getLayer().setBackgroundColor(mOffThumbColor);
        }
    }
}

File: library/src/main/java/com/yy/codex/uikit/UISwitch.kt
package com.yy.codex.uikit

import android.content.Context
import android.graphics.Bitmap
import android.os.Build
import android.support.annotation.RequiresApi
import android.util.AttributeSet
import android.view.View

/**
 * Created by cuiminghui on 2017/1/17.
 */

class UISwitch : UIControl {

    /* initialize */

    constructor(context: Context, view: View) : super(context, view) {}

    constructor(context: Context) : super(context) {}

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {}

    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {}

    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int, defStyleRes: Int) : super(context, attrs, defStyleAttr, defStyleRes) {
    }

    override fun init() {
        super.init()
        initContentView()
        addSubview(contentView)
    }

    override fun intrinsicContentSize(): CGSize {
        return CGSize(52.0, 44.0)
    }

    override fun prepareProps(attrs: AttributeSet) {
        super.prepareProps(attrs)
        val typedArray = context.theme.obtainStyledAttributes(attrs, R.styleable.UISwitch, 0, 0)
        typedArray.getBoolean(R.styleable.UISwitch_switch_on, false)?.let {
            initializeAttributes.put("UISwitch.on", it)
        }
        typedArray.getColor(R.styleable.UISwitch_switch_onThumbColor, -1)?.let {
            if (it != -1) {
                initializeAttributes.put("UISwitch.onThumbColor", UIColor(it))
            }
        }
        typedArray.getColor(R.styleable.UISwitch_switch_onTrackColor, -1)?.let {
            if (it != -1) {
                initializeAttributes.put("UISwitch.onTrackColor", UIColor(it))
            }
        }
        typedArray.getColor(R.styleable.UISwitch_switch_offThumbColor, -1)?.let {
            if (it != -1) {
                initializeAttributes.put("UISwitch.offThumbColor", UIColor(it))
            }
        }
        typedArray.getColor(R.styleable.UISwitch_switch_offTrackColor, -1)?.let {
            if (it != -1) {
                initializeAttributes.put("UISwitch.offTrackColor", UIColor(it))
            }
        }
    }

    override fun resetProps() {
        super.resetProps()
        initializeAttributes?.let {
            (it["UISwitch.on"] as? Boolean)?.let {
                on = it
            }
            (it["UISwitch.onThumbColor"] as? UIColor)?.let {
                onThumbColor = it
            }
            (it["UISwitch.onTrackColor"] as? UIColor)?.let {
                onTrackColor = it
            }
            (it["UISwitch.offThumbColor"] as? UIColor)?.let {
                offThumbColor = it
            }
            (it["UISwitch.offTrackColor"] as? UIColor)?.let {
                offTrackColor = it
            }
        }
    }

    /* appearance */

    private lateinit var contentView: UIView

    private fun initContentView() {
        contentView = UIView(context)
        contentView.constraint = UIConstraint.center()
        contentView.constraint?.width = "52"
        contentView.constraint?.height = "32"
        initOffBackgroundView()
        contentView.addSubview(offBackgroundView)
        initOnBackgroundView()
        contentView.addSubview(onBackgroundView)
        initHandleView()
        contentView.addSubview(handleView)
    }

    private lateinit var offBackgroundView: UIView

    private fun initOffBackgroundView() {
        offBackgroundView = UIView(context)
        offBackgroundView.frame = CGRect(0.0, 0.0, 52.0, 32.0)
        offBackgroundView.alpha = 1.0f
        offBackgroundView.wantsLayer = true
        offBackgroundView.layer.cornerRadius = 16.0
        offBackgroundView.layer.borderWidth = 1.0
        offBackgroundView.layer.borderColor = UIColor(0x00 / 255.0, 0x00 / 255.0, 0x00 / 255.0, 0.1)
        offBackgroundView.layer.backgroundColor = offTrackColor
    }

    private lateinit var onBackgroundView: UIView

    private fun initOnBackgroundView() {
        onBackgroundView = UIView(context)
        onBackgroundView.frame = CGRect(0.0, 0.0, 52.0, 32.0)
        onBackgroundView.alpha = 0.0f
        onBackgroundView.wantsLayer = true
        onBackgroundView.layer.cornerRadius = 16.0
        onBackgroundView.layer.backgroundColor = onTrackColor
    }

    private lateinit var handleView: UIView

    private fun initHandleView() {
        handleView = UIView(context)
        handleView.frame = CGRect(1.0, 1.0, kHandleRadius.toDouble(), kHandleRadius.toDouble())
        handleView.alpha = 1.0f
        handleView.wantsLayer = true
        handleView.layer.shadowX = 1.0
        handleView.layer.shadowY = 1.0
        handleView.layer.shadowRadius = 0.5
        handleView.layer.shadowColor = UIColor(.3, .3, .3, .2)
        handleView.layer.cornerRadius = kHandleRadius / 2.0
        handleView.layer.borderWidth = 0.5
        handleView.layer.borderColor = UIColor(0x00 / 255.0, 0x00 / 255.0, 0x00 / 255.0, 0.15)
        handleView.layer.backgroundColor = UIColor.whiteColor
    }
    
    var on: Boolean = false
        private set

    private var active: Boolean = false

    private var currentAnimation: UIViewAnimation? = null

    var onThumbColor = UIColor.whiteColor
        set(value) {
            field = value
            if (on){
                handleView.layer.backgroundColor = value
            }
        }

    var onTrackColor = UIColor(0x00 / 255.0, 0xe3 / 255.0, 0x64 / 255.0, 1.0)
        set(value) {
            field = value
            onBackgroundView.layer.backgroundColor = value
        }

    var offThumbColor = UIColor.whiteColor
        set(value) {
            field = value
            if (!on){
                handleView.layer.backgroundColor = value
            }
        }

    var offTrackColor = UIColor.whiteColor
        set(value) {
            field = value
            offBackgroundView.layer.backgroundColor = value
        }

    override fun onEvent(event: UIControl.Event) {
        super.onEvent(event)
        when (event) {
            UIControl.Event.TouchDown -> setActive(true)
            UIControl.Event.TouchDragEnter -> setActive(true)
            UIControl.Event.TouchDragExit -> setActive(false)
            UIControl.Event.TouchUpInside -> setOn(!this.on, true)
            UIControl.Event.TouchUpOutside -> setOn(this.on, true)
        }
    }

    fun setOn(value: Boolean, animated: Boolean) {
        on = value
        resetContentView(animated)
        onEvent(Event.ValueChanged)
    }

    private fun resetContentView(animated: Boolean) {
        currentAnimation?.let { it.cancel() }
        if (animated) {
            currentAnimation = UIViewAnimator.springWithBounciness(1.0, 24.0, Runnable {
                if (on) {
                    offBackgroundView.alpha = 0.0f
                    onBackgroundView.alpha = 1.0f
                    onBackgroundView.layer.backgroundColor = onTrackColor
                    handleView.frame = CGRect(frame.width - kHandleRadius.toDouble() - 1.0, 1.0, kHandleRadius.toDouble(), kHandleRadius.toDouble())
                    handleView.layer.backgroundColor = onThumbColor
                } else {
                    offBackgroundView.alpha = 1.0f
                    offBackgroundView.layer.backgroundColor = offTrackColor
                    onBackgroundView.alpha = 0.0f
                    handleView.frame = CGRect(1.0, 1.0, kHandleRadius.toDouble(), kHandleRadius.toDouble())
                    handleView.layer.backgroundColor = offThumbColor
                }
            }, Runnable { active = false })
        }
        else {
            if (on) {
                offBackgroundView.alpha = 0.0f
                onBackgroundView.alpha = 1.0f
                onBackgroundView.layer.backgroundColor = onTrackColor
                handleView.frame = CGRect(frame.width - kHandleRadius.toDouble() - 1.0, 1.0, kHandleRadius.toDouble(), kHandleRadius.toDouble())
                handleView.layer.backgroundColor = onThumbColor
            } else {
                offBackgroundView.alpha = 1.0f
                offBackgroundView.layer.backgroundColor = offTrackColor
                onBackgroundView.alpha = 0.0f
                handleView.frame = CGRect(1.0, 1.0, kHandleRadius.toDouble(), kHandleRadius.toDouble())
                handleView.layer.backgroundColor = offThumbColor
            }
        }
    }

    private fun setActive(isActive: Boolean) {
        this.active = isActive
        cancelAnimation()
        currentAnimation = UIViewAnimator.spring(Runnable {
            val widthExpanded = 8f
            if (isActive) {
                if (on) {
                    handleView.frame = CGRect((21 - widthExpanded).toDouble(), 1.0, (kHandleRadius + widthExpanded).toDouble(), kHandleRadius.toDouble())
                } else {
                    handleView.frame = CGRect(1.0, 1.0, (kHandleRadius + widthExpanded).toDouble(), kHandleRadius.toDouble())
                    offBackgroundView.layer.backgroundColor = offTrackColor.colorWithDarken(0.3)
                }
            } else {
                if (on) {
                    handleView.frame = CGRect(21.0, 1.0, kHandleRadius.toDouble(), kHandleRadius.toDouble())
                } else {
                    handleView.frame = CGRect(1.0, 1.0, kHandleRadius.toDouble(), kHandleRadius.toDouble())
                    offBackgroundView.layer.backgroundColor = offTrackColor
                }
            }
        }, Runnable { active = false })
    }

    private fun cancelAnimation() {
        currentAnimation?.let(UIViewAnimation::cancel)
    }

    companion object {

        private val kHandleRadius = 30.0f

    }

}
